add support for net5.0-<platform>

zkat opened the pull request:
Fixes: https://github.com/NuGet/Home/issues/9240

This PR adds support for `net5.0-<platform>` to `NuGetFramework` and related classes.

This isn't quite ready to merge, but I'd like to start getting reviews about the stuff that's already there. It turns out that I was missing a lot of stuff in https://github.com/NuGet/Home/pull/9382, which I can go back and update the document with, although I'm tempted to just drop that PR altogether in favor of a straight-up implementation.

rrelyea commented on test/NuGet.Core.Tests/NuGet.Frameworks.Test/CompatibilityTests.cs:
```diff
@@ -42,14 +42,18 @@ public class CompatibilityTests
+        [InlineData("net5.0-tvos", "net5.0", true)]
+        [InlineData("net5.0-watchos", "net5.0", true)]
+        [InlineData("net5.0-windows", "net5.0", true)]
+        [InlineData("net5.0", "net5.0-windows", false)]
```
line 53 and 57 are dupes...and build fails.
do you also want to test android <-> windows as false? etc...

rrelyea commented on src/NuGet.Core/NuGet.Frameworks/FrameworkNameProvider.cs:
```diff
@@ -194,6 +207,7 @@ public string GetVersionString(string framework, Version version)
                     versionParts.Pop();
                 }
 
+                // TODO: This comment needs to be updated (maybe the code, too)
```
don't forget this todo.
yes. sorry, i removed two of those cases and didn't fix comment.

rrelyea commented on src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkFactory.cs:
```diff
@@ -330,6 +385,7 @@ private static bool TryParseDeprecatedFramework(string s, out NuGetFramework fra
                 return null;
             }
 
+            // TODO: profile might actually be a Platform. Figure this out.
```
another todo.

rrelyea commented on src/NuGet.Core/NuGet.Packaging/compiler/resources/nuspec.xsd:
```diff
@@ -19,6 +19,8 @@
             </xs:element>
         </xs:sequence>
         <xs:attribute name="targetFramework" type="xs:string" use="optional" />
+        <!-- something like this: (not just a single property. might have multiple dgroups. One per group.) -->
```
todo in nuspec.xsd

rrelyea commented on test/NuGet.Core.Tests/NuGet.Frameworks.Test/CompatibilityTests.cs:
```diff
@@ -42,14 +42,18 @@ public class CompatibilityTests
+        [InlineData("net5.0-android", "net5.0", true)]
+        [InlineData("net5.0-android", "netcoreapp3.1", true)]
+        [InlineData("net5.0-ios", "net5.0", true)]
+        [InlineData("net5.0-macos", "net5.0", true)]
```
>macos [](start = 28, length = 5)

do you have support and test coverage of Platform versions as well yet? coming later for this work? or in a separate task?

rrelyea commented on test/NuGet.Core.Tests/NuGet.Frameworks.Test/CompatibilityTests.cs:
```diff
@@ -42,14 +42,18 @@ public class CompatibilityTests
+        [InlineData("net5.0-android", "netcoreapp3.1", true)]
+        [InlineData("net5.0-ios", "net5.0", true)]
+        [InlineData("net5.0-macos", "net5.0", true)]
+        [InlineData("net5.0-tvos", "net5.0", true)]
```
likely add some compat tests of platforms with platform versions too? (or maybe you have them somewhere else?)

rrelyea commented on test/NuGet.Core.Tests/NuGet.Frameworks.Test/CompatibilityTests.cs:
```diff
@@ -42,14 +42,18 @@ public class CompatibilityTests
+        // net5.0-<osname> is compatible with net5.0 but not the inverse
+        [InlineData("net5.0-android", "net5.0", true)]
+        [InlineData("net5.0-android", "netcoreapp3.1", true)]
+        [InlineData("net5.0-ios", "net5.0", true)]
```
// should catch these problems, likely?
        [InlineData("net5.0-mac os", "net5.0", false)]  // this ignores the space
        [InlineData("net5.0-macos-14", "net5.0", false)]  // this ignores the "-14"
        [InlineData("net5.0-macos-14..0", "net5.0", false)] // not sure what this does.

zkat replied:
right now, all of these parse just fine, oddly enough, because I'm using most of the old parsing code. Should these fail?

rrelyea reviewed (dismissed):
:clock1:

looks very good. found a few minor things so far.

rolfbjarne commented on src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:
```diff
@@ -35,11 +35,17 @@ public static class PlatformIdentifiers
+        {
+            public const string Android = "Android";
+            public const string IOS = "iOS";
+            public const string MacOS = "MacOS";
```
The official name is `macOS` now:

```suggestion
            public const string MacOS = "macOS";
```

rolfbjarne commented on test/NuGet.Core.Tests/NuGet.Frameworks.Test/NuGetFrameworkParseTests.cs:
```diff
@@ -224,6 +225,12 @@ public void NuGetFramework_ProfileName(string folder, string expected)
         [InlineData("netcoreapp3.0", ".NetCoreApp,Version=v3.0")]
+        [InlineData("net5.0-android", ".NETCoreApp,Version=v5.0,Platform=Android")]
+        [InlineData("net5.0-ios14.0", ".NETCoreApp,Version=v5.0,Platform=iOS,PlatformVersion=v14.0")]
+        [InlineData("net5.0-macos10.0", ".NETCoreApp,Version=v5.0,Platform=MacOS,PlatformVersion=v10.0")]
```
Maybe try something like `net5.0-macos10.15.2.3` as well?

cristinamanum commented on src/NuGet.Core/NuGet.Frameworks/CompatibilityProvider.cs:
```diff
@@ -181,18 +182,18 @@ private bool IsCompatibleWithTarget(NuGetFramework target, NuGetFramework candid
-            }
-            else
-            {
-                return NuGetFramework.FrameworkNameComparer.Equals(target, candidate)
```
If the target.IsNetEra == false, previously we used to return line 192 above.

With this change the result is changed to the result below(line 185), is this intentionally? Was a bug there before?

zkat replied:
Can you clarify? I refactored this so the net5 bit was split off into its own section. The logic *should* be about the same, except for the Platform parts.

cristinamanum replied:
I misunderstood the change. I read it again and all good.

cristinamanum commented on src/NuGet.Core/NuGet.Frameworks/FrameworkNameProvider.cs:
```diff
@@ -124,6 +127,11 @@ public bool TryGetProfile(string frameworkIdentifier, string profileShortName, o
             return TryConvertOrNormalize(profileShortName, _profileShortToLong, _profilesToShortName, out profile);
         }
 
+        public bool TryGetPlatform(string frameworkIdentifier, string platformShortName, out string platform)
```
The "frameworkIdentifier" is not used.

cristinamanum replied:
Will the "platform" be a long name? Maybe add method comments?

cristinamanum commented on src/NuGet.Core/NuGet.Frameworks/FrameworkNameProvider.cs:
```diff
@@ -134,6 +142,11 @@ public bool TryGetShortProfile(string frameworkIdentifier, string profile, out s
             return TryConvertOrNormalize(profile, _profilesToShortName, _profileShortToLong, out profileShortName);
         }
 
+        public bool TryGetShortPlatform(string frameworkIdentifier, string platform, out string platformShortName)
```
The "frameworkIdentifier" is not used.

cristinamanum commented on src/NuGet.Core/NuGet.Frameworks/NuGetFramework.cs:
```diff
@@ -57,8 +57,34 @@ public NuGetFramework(string frameworkIdentifier, Version frameworkVersion, stri
+
+            IsNet5Era = (_frameworkVersion.Major >= Version5 && StringComparer.OrdinalIgnoreCase.Equals(FrameworkConstants.FrameworkIdentifiers.NetCoreApp, _frameworkIdentifier));
+
+            _frameworkProfile = profile ?? string.Empty;
```
Why not throwing if null profile?

zkat replied:
that would be a breaking change, so I can't do that one. :(

cristinamanum commented on src/NuGet.Core/NuGet.Frameworks/NuGetFramework.cs:
```diff
@@ -57,8 +57,34 @@ public NuGetFramework(string frameworkIdentifier, Version frameworkVersion, stri
+            {
+                throw new ArgumentNullException("frameworkVersion");
+            }
+
```
Why not argument null check for the platform and platformVersion?

cristinamanum commented on src/NuGet.Core/NuGet.Frameworks/NuGetFramework.cs:
```diff
@@ -57,8 +57,34 @@ public NuGetFramework(string frameworkIdentifier, Version frameworkVersion, stri
+            PlatformVersion = FrameworkConstants.EmptyVersion;
+        }
+
+        public NuGetFramework(string frameworkIdentifier, Version frameworkVersion, string platform, Version platformVersion)
```
The logic for this constructor depends on the IsNet5Era. I think that method comments would be useful for the consumers.

nkolev92 commented on src/NuGet.Core/NuGet.Frameworks/NuGetFramework.cs:
```diff
@@ -57,8 +60,47 @@ public NuGetFramework(string frameworkIdentifier, Version frameworkVersion, stri
+        {
+            if (frameworkIdentifier == null)
+            {
+                throw new ArgumentNullException("frameworkIdentifier");
```
style: Consider using nameof(frameworkIdentifier)

zkat replied:
I don't think we should change this because the public API has been hard-coded to this string and it might be weird for an overload to have different behavior.

nkolev92 replied:
Renaming a parameter is technially a breaking change, so we shouldn't do that. 

Not sure how I feel about that motivation in particular. 

Throwing an argument null exception for a parameter name that's potentially wrong is also bad.

zkat replied:
I tried to do this and... uhh?

![image](https://user-images.githubusercontent.com/17535/80138345-86e59d80-8559-11ea-9ebc-a308da3d682d.png)

nkolev92 replied:
Waaatt...NuGetFramework is still targetting net40 :D 
Not super relevant in the grand scheme of things. Let's just move on :)

nkolev92 commented on src/NuGet.Core/NuGet.Frameworks/NuGetFramework.cs:
```diff
@@ -57,8 +60,47 @@ public NuGetFramework(string frameworkIdentifier, Version frameworkVersion, stri
+
+            if (frameworkVersion == null)
+            {
+                throw new ArgumentNullException("frameworkVersion");
```
similar nit/style feedback applies to the rest.

zkat replied:
see above comment -- I don't think we can use this here.

nkolev92 commented on src/NuGet.Core/NuGet.Frameworks/NuGetFramework.cs:
```diff
@@ -120,10 +188,18 @@ public string GetDotNetFrameworkName(IFrameworkNameProvider mappings)
                 parts.Add(string.Format(CultureInfo.InvariantCulture, "Version=v{0}", GetDisplayVersion(framework.Version)));
 
-                if (!string.IsNullOrEmpty(framework.Profile))
+                if (!IsNet5Era && !string.IsNullOrEmpty(framework.Profile))
```
Does this really matter?
Since we are not using profiles anymore, this should not be relevant.

zkat replied:
I'm just being doubleplus sure to express that Profiles are just not a thing in net5 in any way. Would you like me to remove this?

nkolev92 replied:
Your call. 
We have not tried to do restrictions like this in the past because it complicates things. 
Granted right now we're being too permissive.

My main concern is readability though (given that this has no actual implementation impact because of how these properties are being set right now), so maybe a judgment call (or wait for someone else to complain, lol).

nkolev92 commented on src/NuGet.Core/NuGet.Frameworks/NuGetFramework.cs:
```diff
@@ -120,10 +188,18 @@ public string GetDotNetFrameworkName(IFrameworkNameProvider mappings)
                 {
                     parts.Add(string.Format(CultureInfo.InvariantCulture, "Profile={0}", framework.Profile));
                 }
+                if (IsNet5Era && !string.IsNullOrEmpty(framework.Platform))
```
Similarly, I'm not sure there's much value in restricting platform to .NET 5 era only. 
Don't feel strongly about this (yet :P ), but mostly curious if the added perspective makes you reconsider this.

nkolev92 replied:
Especially given that this is immutable and there's absolutely no way to create a framework object with both a platform and profile at this point.

nkolev92 commented on src/NuGet.Core/NuGet.Frameworks/NuGetFramework.cs:
```diff
@@ -120,10 +188,18 @@ public string GetDotNetFrameworkName(IFrameworkNameProvider mappings)
+                    parts.Add(string.Format(CultureInfo.InvariantCulture, "Platform={0}", framework.Platform));
+                    if (framework.PlatformVersion != FrameworkConstants.EmptyVersion)
+                    {
+                        parts.Add(string.Format(CultureInfo.InvariantCulture, "PlatformVersion=v{0}", GetDisplayVersion(framework.PlatformVersion)));
```
did we vet this with @terrajobst @dsplaisted @mhutch?

dsplaisted replied:
I was thinking that the platform information would not round-trip into the .NET TargetFrameworkMoniker.  Is there any need to represent the platform info there?

nkolev92 replied:
This is what we write in the nuspec. 
The platform & version are a part of the more complete framework identifier.

dsplaisted replied:
If it just goes in the nuspec then it's more or less an implementation detail of NuGet.  

- The string does surface in some cases (I think NuGet.org uses it to some degree)
- We had talked about changing the format of this information in the nuspec to use the short framework name.  This would avoid a net5.0 package still showing .NETCoreApp as the target framework identifier
- Is the nuspec the only place this shows up?  Are there other (possibly non-NuGet) consumers of the string in this format?

zkat replied:
Nope, this logic hasn't been vetted. This might be worth discussing, it seems?

nkolev92 replied:
We write it in the nuspec.
We show it in the PM UI, nuget.org etc.

This is the primary "pivot" for packages though. 

So everyone should be able to see a dependency group is specific to the IOS platform.

Nirmal4G replied:
I've suggested in the design issue to use `TargetPlatform*` properties separately without linking the `TargetFramework*` property, in which case nuget adds `NuGetPlatform` as opposed to `NuGetFramework` and similar classes/property names. If that is too much then use profiles, its a lot simpler than adding `Platform` concept to nuget.

zkat replied:
This logic has been updated to follow https://github.com/dotnet/designs/pull/114

nkolev92 commented on src/NuGet.Core/NuGet.Frameworks/NuGetFramework.cs:
```diff
@@ -218,6 +294,25 @@ public virtual string GetShortFolderName(IFrameworkNameProvider mappings)
                             framework.DotNetFrameworkName));
                     }
                 }
+                else if (IsNet5Era)
```
The code inside this block looks good, I'm just wandering if we are benefiting from the IsNet5Era check.

nkolev92 commented on src/NuGet.Core/NuGet.Packaging/compiler/resources/nuspec.xsd:
```diff
@@ -51,7 +51,7 @@
       <xs:attribute name="targetFramework" type="xs:string" use="required" />
     </xs:complexType>
-  
+
```
revert unnecessary changes, these can be cleaned up later.

zkat replied:
oops! I thought I'd gotten rid of this lol

nkolev92 commented on test/NuGet.Core.Tests/NuGet.Frameworks.Test/NuGetFrameworkParseTests.cs:
```diff
@@ -224,6 +225,13 @@ public void NuGetFramework_ProfileName(string folder, string expected)
         [InlineData("netcoreapp1.5", ".NETCoreApp,Version=v1.5")]
         [InlineData("netcoreapp2.0", ".NetCoreApp,Version=v2.0")]
         [InlineData("netcoreapp3.0", ".NetCoreApp,Version=v3.0")]
+        [InlineData("net5.0-android", ".NETCoreApp,Version=v5.0,Platform=Android")]
```
I think I'd expect PlatformVersion=v0.0 similar to what "netcoreapp" would do for the framework version. 

Platform and Platform version *always* go together.

nkolev92 replied:
actually...this is reversed from most of our tests...so the input is "Platform=Android" haha. 

I'd also test PlatformVersion=v0.0 in that case.

zkat replied:
obsolete -- this is no longer the format.

nkolev92 commented on test/NuGet.Core.Tests/NuGet.Frameworks.Test/CompatibilityTests.cs:
```diff
@@ -42,15 +42,36 @@ public class CompatibilityTests
         [InlineData("net5.0", "net5.0-windows", false)]
+        [InlineData("net5.0-mac os", "net5.0", false)]
+        [InlineData("net5.0-macos-14", "net5.0", false)]
+        [InlineData("net5.0-macos-14..0", "net5.0", false)]
```
How does this parse?
What are the platform/platform version?

zkat replied:
These parse to Unsupported. I'll document that.

nkolev92 reviewed (commented):
Looks good so far. 

A few questions/suggestions. 
No complaints on the public API part.

zkat committed:
add support for net5.0-<platform>

Fixes: https://github.com/NuGet/Home/issues/9240

add TFI and TFV arguments + return null for GetNearest

stop verifying platform names

let bad frameworkname parses fail explosively

fix tests

more PR feedback fixes

normalize platform shortnames and suppress resulting warning

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileTarget.cs:
```diff
@@ -15,7 +15,7 @@ public class LockFileTarget : IEquatable<LockFileTarget>
         public string RuntimeIdentifier { get; set; }
 
-        public string Name => TargetFramework + (string.IsNullOrEmpty(RuntimeIdentifier) ? "" : "/" + RuntimeIdentifier);
+        public string Name => TargetFramework.GetShortFolderName() + (string.IsNullOrEmpty(RuntimeIdentifier) ? "" : "/" + RuntimeIdentifier);
```
Do we have roundtrip tests for this?

zkat replied:
Note: The spec was updated, so we no longer have assets file changes.

nkolev92 reviewed (commented):
LGTM.

Can we add some tests for the assets file change?

I assume the pack change will be a new PR?

zkat commented:
@nkolev92 do you still feel we need tests, now that the assets file changes aren't there anymore? I'm not sure what additional stuff to test, right now.

nkolev92 commented:
@zkat I'll look at this later today.

nkolev92 reviewed (commented):
I think we don't need more tests for the assets file and lock file given the current design, however DotnetFrameworkName is caused by extensibility APIs in VS and those end up using the FrameworkName type from System.Runtime.Versioning: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsFrameworkCompatibility.cs

Those APIs might throw, so I think we should validate their behavior.

zivkan commented on src/NuGet.Core/NuGet.Frameworks/FrameworkNameProvider.cs:
```diff
@@ -124,6 +127,11 @@ public bool TryGetProfile(string frameworkIdentifier, string profileShortName, o
             return TryConvertOrNormalize(profileShortName, _profileShortToLong, _profilesToShortName, out profile);
         }
 
+        public bool TryGetPlatform(string platformShortName, out string platformIdentifier)
```
Since this is public, once it ships, it can basically never be changed. The comment for `FrameworkPlatformIdentifiers` in `FrameworkConstants.cs` says that it's the .NET 5 era platform name. And if .NET 6 or later add new (or redefine?) a platform, those platforms will not be valid for all TFMs (in fact, none of the platforms are valid for .NET Framework or .NET Standard, or .NET Core App < 5.0).

Therefore, shouldn't this API have more inputs to specific TFI and TFV?

zkat replied:
we can do that, I think. I'll give it a shot and see how it affects things.

zivkan commented on src/NuGet.Core/NuGet.Frameworks/FrameworkNameProvider.cs:
```diff
@@ -134,6 +142,11 @@ public bool TryGetShortProfile(string frameworkIdentifier, string profile, out s
             return TryConvertOrNormalize(profile, _profilesToShortName, _profileShortToLong, out profileShortName);
         }
 
+        public bool TryGetShortPlatform(string platformIdentifier, out string platformShortName)
```
Like the other comment, I fear this API might not be a good design for future expansion

zivkan commented on src/NuGet.Core/NuGet.Frameworks/def/IFrameworkMappings.cs:
```diff
@@ -34,6 +34,11 @@ interface IFrameworkMappings
+        /// <summary>
+        /// Ex: iOS &#8210;&gt; ios
+        /// </summary>
+        IEnumerable<KeyValuePair<string, string>> PlatformShortNames { get; }
```
Doesn't platform short names depend on the TFI & TFV?

zivkan commented on src/NuGet.Core/NuGet.Frameworks/def/IFrameworkNameProvider.cs:
```diff
@@ -33,6 +33,16 @@ interface IFrameworkNameProvider
+        /// <summary>
+        /// Get the official platform name from the short name.
+        /// </summary>
+        bool TryGetPlatform(string platformShortName, out string platform);
```
TFI & TFV?

zivkan commented on src/NuGet.Core/NuGet.Frameworks/def/IFrameworkNameProvider.cs:
```diff
@@ -33,6 +33,16 @@ interface IFrameworkNameProvider
+        /// <summary>
+        /// Get the official platform name from the short name.
+        /// </summary>
+        bool TryGetShortPlatform(string platform, out string platformShortName);
```
TFI & TFV?

Nirmal4G commented:
What about adding Platform concept to NuGet? That may be a lot of work but will generalize the work happening for .NET 5, C++ `ProjectReference`. We could introduce `NuGet.Platforms` lib for holding the platforms logic similar to frameworks in `NuGet.Frameworks`!

Is that too much?

nkolev92 commented:
@Nirmal4G In the current design, the platform and version are sort of a sub-concept of the larger target framework concept.

The pivot for packages remains the target framework and target framework only.
Even in the .NET 5 and later world.

Nirmal4G commented:
> The pivot for packages remains the target framework and target framework only.

We say that now. but didn't we had similar issues in the past with framework profiles and Xamarin TFMs?
_I can see the history is repeating itself but with different parameters_ 😉

Then, this means that the `-{platfrom}{version}` is similar to profile?

> once it ships, it can basically never be changed

Diverging the code like this could introduce a lot of maintainability problems in the future, IMHO. It is better to treat them as separate concepts, It'll also align better, with RIDs.

zivkan commented:
> Diverging the code like this could introduce a lot of maintainability problems in the future

I don't understand what you mean by this.

My feedback was in response to the fact we're implement this .NET design: https://github.com/dotnet/designs/blob/master/accepted/2020/net5/net5.md

It lists several platforms which will be supported in .NET 5, and my understanding of the implementation of NuGetFramework in this PR is such that `NuGetFramework.Parse("net5.0-zivkan")` will return `FrameworkConstants.Unsupported` (I'm guessing the type name, but my point is it doesn't return a .NET 5 TFM with the platform set to "zivkan", it returns the unsupported framework - another API design choice I hate, but that's what we live with).

If in the future `zivkan` becomes a real platform, it doesn't change the fact that `net5.0-zivkan` is not valid, and never will be valid. People creating NuGet multi-targeting NuGet packages can target `net123.0-zivkan`, but never `net5.0-zivkan`.

The APIs added here to get platform names is incapable of understanding the subtlety of when a platform is supported in some target framework versions, but not others.

nkolev92 commented:
@Nirmal4G 
> We say that now. but didn't we had similar issues in the past with framework profiles and Xamarin TFMs?

That's exactly why there is a platform version as part of the framework and TargetPlatformMinVersion. 

I'd provide such feedback in the dotnet/designs repo.

zkat commented:
@nkolev92 @zivkan I believe I made the changes as requested. Can you let me know if this is in line with what you had in mind? Happy to change things as needed.

ericstj commented:
> I can see the history is repeating itself but with different parameters 😉

I agree completely.  I would like to refcount the concern that you are baking in the policy around valid platform names into NuGet here.  

Where in the design does it say that ***NuGet*** must be the one that validates these names?  Wouldn't it be better if you just let the **SDK** do the validation.  The SDK must change to support a new platform, that's clear: since it needs to provide reference assemblies and build logic for that platform.

I don't see why NuGet *must* change in order to support a new platform.  I think this creates more moving pieces that will make it harder to react to new platforms and require us to do things like update VS in order to add new platforms.

@terrajobst @mhutch

zivkan commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsFrameworkCompatibility.cs:
```diff
@@ -91,7 +91,17 @@ public FrameworkName GetNearest(FrameworkName targetFramework, IEnumerable<Frame
+            }
+            catch (System.ArgumentException)
+            {
+                // GetNearest isn't supported for net5-era frameworks, since
+                // their identifiers don't losslessly translate to
+                // `FrameworkName`s.
+                return null;
```
Aren't all net5.0-* TFMs the .NETCoreApp framework name?

Even if it doesn't allow us to distinguish between "net5.0" and "net5.0-android", I'm scared that random VS components will break with .NET 5 projects if we return null here.

zkat replied:
Shall we just let this crash, then?

zivkan commented on src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkFactory.cs:
```diff
@@ -196,18 +205,31 @@ public static NuGetFramework ParseFolder(string folderName, IFrameworkNameProvid
-                                    bool validProfile = FrameworkConstants.FrameworkProfiles.Contains(profileShort);
-                                    if (validProfile)
+                                    // Find a platform version if it exists and yank it out
+                                    var platformChars = profileShort.ToCharArray();
```
`string` can be used like a `char[]`, so this line causes needless memory allocations.

zivkan commented on src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:
```diff
@@ -35,11 +35,17 @@ public static class PlatformIdentifiers
-                { };
+        internal static class FrameworkPlatformIdentifiers
+        {
+            public const string Android = "Android";
+            public const string IOS = "iOS";
+            public const string MacOS = "macOS";
+            public const string TvOS = "tvOS";
+            public const string WatchOS = "watchOS";
+            public const string Windows = "Windows";
```
Can we make this all `internal` please, to match the class. It reduces the change of things accidentally becoming public, even though semantically there's no difference since the class is internal.

zivkan commented:
> > I can see the history is repeating itself but with different parameters 😉
> 
> I agree completely. I would like to refcount the concern that you are baking in the policy around valid platform names into NuGet here.

I also agree. I'd prefer if my team is not responsible for maintaining a whitelist of known platforms per TFI/TFV pair.

From what I saw in the code, it's needed to be able to map between the correct capitalization of the TargetPlatformIdentifier and the "short name". It also theoretically allows us to support the TPI being a different string to the short name, but so far the platforms supported just have the equivalent of `shortName = TargetPlatformIdentifier.ToLowerInvarient()`.

If everyone (NuGet, SDK, and anyone else who writes code comparing TargetPlatformIdentifier strings) treats TPI as case insenstive, I hope we can move this out of NuGet and into the SDK. But if there is a need for case sensitivity, then I don't have any ideas because NuGet will need a way to convert from lower case short name to correct case identifier.

mhutch commented:
NuGet needs to be aware of the relationship between existing TFMs and the corresponding net5+ platforms, but yes, I don't think it should necessarily validate additional/unknown platforms

dsplaisted commented:
I agree with @mhutch.  Let's try to have NuGet allow any string for the TargetPlatformIdentifier, and only hard-code the platforms where it needs special rules for compatibility relationships.  The SDK will be responsible for erroring out for unsupported platforms.  That's captured here: https://github.com/dotnet/sdk/issues/11232

nkolev92 commented:
I am ok with NuGet allowing any `TargetPlatformIdentifier` values. It's more flexible this way.

We should just make sure everyone normalizes with ToLower (that's what NuGet will do for the lib/net5-tpitpv). 

> Where in the design does it say that NuGet must be the one that validates these names?

@ericstj 

fwiw, I distinctly remember @rrelyea asking whether NuGet should validate the platform names and the answer back then was yes. 
It was also captured in:
https://github.com/dotnet/designs/blob/master/accepted/2020/net5/net5.md#tfms-are-a-closed-set

We should update the doc to reflect this change: 

@terrajobst wrote at that time

> We may open this up in the future, but for now the consensus was that we'd prefer to have the specific list of names that are expanded by NuGet.

@Nirmal4G

> Then, this means that the -{platfrom}{version} is similar to profile?

It's pretty close, yes.

cristinamanum commented on src/NuGet.Core/NuGet.Frameworks/FrameworkNameProvider.cs:
```diff
@@ -134,6 +137,20 @@ public bool TryGetShortProfile(string frameworkIdentifier, string profile, out s
             return TryConvertOrNormalize(profile, _profilesToShortName, _profileShortToLong, out profileShortName);
         }
 
+        public bool TryGetPlatform(string frameworkIdentifier, Version frameworkVersion, string platformIdentifier, out string platformShortName)
```
Are there null checks needed?

cristinamanum replied:
The platformIdentifier is not used for an identification is this intended?

cristinamanum replied:
Do we need UnitTests for this method?

cristinamanum commented on src/NuGet.Core/NuGet.Frameworks/def/IFrameworkNameProvider.cs:
```diff
@@ -33,6 +33,11 @@ interface IFrameworkNameProvider
         bool TryGetShortProfile(string frameworkIdentifier, string profile, out string profileShortName);
 
+        /// <summary>
+        /// Get the official platform name from the short name.
```
Isn't it the other way: "Get the official platform short name from the platform name" ? The out argument is `platformShortName`

erdembayar commented on src/NuGet.Core/NuGet.Frameworks/FrameworkNameProvider.cs:
```diff
@@ -73,6 +74,8 @@ public FrameworkNameProvider(IEnumerable<IFrameworkMappings> mappings, IEnumerab
             _profileShortToLong = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            _platformToShortName = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            _platformShortToLong = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
             _portableFrameworks = new Dictionary<int, HashSet<NuGetFramework>>();
```
Added 2 dictionaries( _platformToShortName,  _platformShortToLong ) are initialized values, but never used.
Why is that?

zkat replied:
oops, it's a leftover from a previous iteration.

erdembayar commented on src/NuGet.Core/NuGet.Frameworks/CompatibilityProvider.cs:
```diff
@@ -48,8 +48,9 @@ public bool IsCompatible(NuGetFramework target, NuGetFramework candidate)
-            bool? result = _cache.GetOrAdd(cacheKey, (Func<CompatibilityCacheKey, bool>)((key) 
-                => { return IsCompatibleCore(target, candidate) == true; }));
+            bool? result = _cache.GetOrAdd(cacheKey, (Func<CompatibilityCacheKey, bool>)((key)
+                =>
```
Can we simplify this part like `bool? result = _cache.GetOrAdd(cacheKey, IsCompatibleCore(target, candidate) == true);`
Otherwise very hard to read and understand.

zivkan replied:
This looks to me like a re-formatting of existing code, but its functionality hasn't changed at all. I'm all for simplifying syntax, but I consider it out of scope for the PR (I wouldn't complain if it's improved, but I wouldn't hold up the PR because of it)

erdembayar commented on src/NuGet.Core/NuGet.Frameworks/NuGetFramework.cs:
```diff
@@ -77,6 +119,32 @@ public Version Version
+        /// True if the platform is non-empty
+        /// </summary>
+        public bool HasPlatform
+        {
```
Can we use more modern syntax here? `public bool HasPlatform => !string.IsNullOrEmpty(Platform);`

zkat replied:
I'd rather keep things consistent with the rest of the file, and I think it's out of scope to change all the other ones that work this way.

erdembayar commented on src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkFactory.cs:
```diff
@@ -92,18 +92,19 @@ public static NuGetFramework ParseFrameworkName(string frameworkName, IFramework
             // if the first part is a special framework, ignore the rest
             if (!TryParseSpecialFramework(parts[0], out result))
             {
```
Just in case add `parts.Any()` if expecting at least 1 item in array which is parts[0] for not to get index out of range exception.
`if (parts.Any() && !TryParseSpecialFramework(parts[0], out result))`

zivkan replied:
When you know the data type, using `.Length > 0` or `.Count > 0` is a lot faster than using LINQ to do the same. Of course, this one thing is such a small part of the overall execution that in customer experience % terms it'll be just about zero, but I personally encourage people to avoid LINQ where possible.

erdembayar commented on src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkFactory.cs:
```diff
@@ -196,18 +205,30 @@ public static NuGetFramework ParseFolder(string folderName, IFrameworkNameProvid
-                                    bool validProfile = FrameworkConstants.FrameworkProfiles.Contains(profileShort);
-                                    if (validProfile)
+                                    // Find a platform version if it exists and yank it out
+                                    var platformChars = profileShort;
```
No need this extra `platformChars` variable here, keep using `profileShort` unless we do `profileShort.ToCharArray()`, but most likely no need;

nkolev92 replied:
This improves the readability because the 3rd part can be either profile or platform depending on the context. 

I'd expect both of these to compile to the same thing anyways.

dsplaisted commented:
> fwiw, I distinctly remember @rrelyea asking whether NuGet should validate the platform names and the answer back then was yes.
> It was also captured in:
> https://github.com/dotnet/designs/blob/master/accepted/2020/net5/net5.md#tfms-are-a-closed-set
> 
> We should update the doc to reflect this change:

@nkolev92 I've sent a designs PR to cover this: https://github.com/dotnet/designs/pull/130

zivkan commented on src/NuGet.Core/NuGet.Frameworks/FrameworkNameProvider.cs:
```diff
@@ -194,8 +211,9 @@ public string GetVersionString(string framework, Version version)
 
-                // Always use decimals and 2+ digits for dotnet, netstandard, netstandardapp,
-                // netcoreapp, or if any parts of the version are over 9 we need to use decimals
+                // Always use decimals and 2+ digits for netstandard and
+                // netcoreapp, or if any parts of the version are over 9
+                // we need to use decimals
```
I wouldn't have noticed if you didn't reformat this comment, but I don't believe this is correct any longer.  net5.0 and above (net5Era) should also be using ".0" at the end.

Well, technically `net5` will work just fine, but we want to make sure any time we (or the SDK, or anything else calling NuGet's APIs) write the short name, we always use the ".0" version, so customers are used to that syntax for when "10.0" comes along. But I'm not sure if this is a non-net5era codepath, so maybe it's ok.

zkat replied:
net5.0 is netcoreapp everywhere internally, so the comment is accurate, and the behavior is correct.

zivkan commented on test/NuGet.Clients.Tests/NuGet.VisualStudio.Implementation.Test/Extensibility/VsFrameworkCompatibilityTests.cs:
```diff
@@ -52,6 +52,24 @@ public void VsFrameworkCompatibility_GetNearestRejectsNullFallbackFrameworks()
         }
 
+        [Fact]
+        public void VsFrameworkCompatibility_GetNearestNet5EraFails()
```
Why do we want this to fail?  Isn't there a risk that a bunch of VS components and extensions will needlessly fail with net5.0 projects?

zkat replied:
This API deals directly with `FrameworkName` objects, which explicitly do not support `platform`. My previous iteration on this patch actually made this return null, but some feedback from @nkolev92 convinced me to change it so the error would be LOUD

zivkan replied:
ok, works for me.

nkolev92 replied:
I guess the one concern I'd have here is what are the consumers of this API supposed to do? 

At this point we're not giving them a workaround.. 

I'm not sure who the consumers of this API are. 

https://github.com/search?p=3&q=IVsFrameworkCompatibility&type=Code shows only project system. 

You can answer this question in a future PR if it turns out we need a change :)

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsFrameworkCompatibility.cs:
```diff
@@ -1,4 +1,4 @@
-﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Copyright (c) .NET Foundation. All rights reserved.
```
Can you please revert this. 
Looks like a BOM only change.

nkolev92 commented on src/NuGet.Core/NuGet.Frameworks/FrameworkNameProvider.cs:
```diff
@@ -134,6 +135,20 @@ public bool TryGetShortProfile(string frameworkIdentifier, string profile, out s
+        {
+            if (StringComparer.OrdinalIgnoreCase.Equals(frameworkIdentifier, FrameworkConstants.FrameworkIdentifiers.NetCoreApp) && frameworkVersion.Major >= 5)
+            {
+                platformShortName = platformIdentifier;
```
I think we should normalize somewhere. 

Specifically, GetShortFolderName of

`net5.0-ios` and `net5.0-ioS` should normalize to the same lowercase string. 

At least that my interpretation of the below discussion.

nkolev92 reviewed (approved):
Looks great overall. 

It's been a long journey, where the bulk of work was agreeing to the right design :)
Great job!

2 smaller comments left. First one is more important and I feel more confident about it. 
2nd one is up to you and based on your analysis. 


* I think GetShortFolderName should normalize by providing the same casing. See: https://nugettoolsdev.azurewebsites.net/5.5.0-preview.2.6382/parse-framework?framework=NEtcoreapp1.0 and https://nugettoolsdev.azurewebsites.net/5.5.0-preview.2.6382/parse-framework?framework=NEtcoreapp1.0. Maybe I missed where it's getting normalized, so let me know.

* What's the guidance for customers that can't use the extensibility framework API. Will anyone of note be affected? Do we need an issue tracking that.

zkat commented:
@nkolev92 I've created an issue for it: https://github.com/NuGet/Home/issues/9650
